networkx/generators/joint_degree_seq.py

```diff
@@ -0,0 +1,286 @@
+        if (v not in w_prime_neighbs) and (v!=w_prime):
+            t = v
+            break;
+    
+    # removes (w,t), add (w_prime,t)  and update data structures
+    G.remove_edge(w, t)
+    G.add_edge(w_prime, t)
+    h_node_residual[w] += 1                            
+    h_node_residual[w_prime] -= 1                        
+    if h_node_residual[w_prime] == 0:
+        unsat.remove(w_prime)
+
+        
+def joint_degree_model(nkk,seed=None):
+    """Return a random simple graph with the given joint degree dict (nkk).
+
+    Parameters
+    ----------
+    nkk :  dictionary of dictionary of integers
+        joint degree dictionary. for nodes of degree k (first level of dict) and
+        nodes of degree l (second level of dict) describes the number of edges        
+    seed : hashable object, optional
+        Seed for random number generator.
+
+    Returns
+    -------
+    G : Graph
+        A graph with the specified joint degree dict
+
+    Raises
+    ------
+    NetworkXError
+        If nkk, the joint degree dictionary, is not realizable.
+
+
+    Notes
+    -----
+    In each iteration of the "while loop" the algorithm picks two disconnected nodes v and w,
+    of degree k and l correspondingly,  for which nkk[k][l] has not reached its target yet 
+    i.e. (for given k,l): n_edges_add < nkk[k][] . It then adds edge (v,w) and always increases
+    the number of edges in graph G by one.  
+    
+    The intelligence of the algorithm lies in the fact that  it is always  possible 
+    to add an edge between disconnected nodes v and w, for which nkk[degree(v)][degree(w)] 
+    has not reached its target, even if one or both nodes do not have free stubs. If either node 
+    v or w   does not have a free stub, we perform a "neighbor switch", an edge rewiring move that 
+    releases a free stub while keeping nkk the same.
+    
+    The algorithm continues for E (number of edges in the graph) iterations of the "while loop",
+    at the which point all entries of the given nkk[k][l] have reached their
+    target values and the construction is complete.        
+
+    References
+    ----------
+    [1] M. Gjoka, B. Tillman, A. Markopoulou, "Construction of Simple Graphs 
+       with a Target Joint Degree Matrix and Beyond", IEEE Infocom, 2015.
+
+    Examples
+    --------
+    >>> import networkx as nx
+    >>> joint_degree_dict = {1: {4: 1},
+    ...                      2: {2: 2, 3: 2, 4: 2},
+    ...                      3: {2: 2, 4: 1},
+    ...                      4: {1: 1, 2: 2, 3: 1}}
+    >>> G=nx.joint_degree_model(joint_degree_dict)
+    >>>
+
+    """
+    
+    if not is_valid_joint_degree(nkk):
+        msg = 'Input joint degree dict not realizable as a simple graph'
+        raise nx.NetworkXError(msg)
+
+    if not seed is None:
+        random.seed(seed)
+
+    # compute degree vector from joint degree nkk
+    nk = { deg:(sum(val for val in nkk[deg].itervalues())/deg) for deg in nkk if deg>0 }
```

Comment:
This should maybe be

    nk = {source: sum(target.values()) // source for source, target in nkk.items() if source > 0}

but with better names than `source` and `target`. Notice the integer division, since you are guaranteed integers by invoking `is_valid_joint_degree()` above.